File: java/LongestCommonPrefix14.java
public class LongestCommonPrefix14 {
    public String longestCommonPrefix(String[] strs) {
        if (strs.length == 0)
            return "";
        if (strs.length == 1)
            return strs[0];
        String f = strs[0];
        for (int i = 0; i < f.length(); i++) {
            char c = f.charAt(i);

            for (int j = 1; j < strs.length; j++) {
                if (i >= strs[j].length() || c != strs[j].charAt(i)) {
                    return f.substring(0, i);
                }
            }
        }
        return f;
    }
}

File: rust/src/leetcode/longest_common_prefix_14.rs
use crate::leetcode::Solution;

impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        let mut index = 0;
        'o: loop {
            let mut s = None;
            for x in strs.iter() {
                match x.get(index..index + 1) {
                    Some(c) if s.is_none() => {
                        s = Some(c);
                    }
                    Some(c) if s.unwrap() != c => {
                        break 'o;
                    }
                    Some(_c) => {}
                    None => {
                        break 'o;
                    }
                }
            }
            index += 1;
        }

        strs[0].get(0..index).unwrap().to_string()
    }
}

#[cfg(test)]
mod tests {
    use crate::leetcode::Solution;

    #[test]
    fn test_longest_common_prefix() {
        let vec = vec![
            String::from("flower"),
            String::from("flow"),
            String::from("flight"),
        ];

        let s = Solution::longest_common_prefix(vec);
        assert_eq!("fl", s);

        let vec = vec![
            String::from("dog"),
            String::from("racecar"),
            String::from("car"),
        ];

        let s = Solution::longest_common_prefix(vec);
        assert_eq!("", s);
    }
}
